File: v1/src/main/java/com/api/v1/common/Result.java
package com.api.v1.common;

public final class Result<T> {

    private ResultStatus status;
    private String message;
    private T body;

    private Result() {}

    private Result(ResultStatus status, String message, T body) {
        this.status = status;
        this.message = message;
        this.body = body;
    }

    private Result(ResultStatus status, String message) {
        this.status = status;
        this.message = message;
    }

    public static Result<Void> empty() {
        return new Result<>();
    }

    public static <T>  Result<T> success(T body) {
        return new Result<>(
                ResultStatus.SUCCESSFUL,
                "Successful operation.",
                body
        );
    }

    public static <T>  Result<T> created(T body) {
        return new Result<>(
                ResultStatus.SUCCESSFUL,
                "Resource created.",
                body
        );
    }

    public static <T>  Result<T> error(String message) {
        return new Result<>(
                ResultStatus.ERROR,
                message
        );
    }

    public ResultStatus status() {
        return status;
    }

    public String message() {
        return message;
    }

    public T body() {
        return body;
    }
}

File: v2/src/main/kotlin/com/api/v2/common/Result.kt
package com.api.v2.common

class Result<T> private constructor(
    val status: ResultStatus,
    val message: String,
    val body: T?
) {

    companion object {
        fun <T> success(body: T): Result<T> {
            return Result(
                ResultStatus.SUCCESSFUL,
                "Successful operation.",
                body
            )
        }

        fun <T> created(body: T): Result<T> {
            return Result(
                ResultStatus.SUCCESSFUL,
                "Resource created.",
                body
            )
        }

        fun <T> error(message: String): Result<T> {
            return Result(
                ResultStatus.ERROR,
                message,
                null
            )
        }
    }
}